Split WebInspectorUIProxy into separate interfaces for frontend and backend

jon-b-apple opened the pull request:
#### d52237d4477592061954b6e83c0ee664fba9f4be
<pre>
Split WebInspectorUIProxy into separate interfaces for frontend and backend
<a href="https://rdar.apple.com/155312814">rdar://155312814</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295579">https://bugs.webkit.org/show_bug.cgi?id=295579</a>

Reviewed by BJ Burg.

First set of changes to split the backend-accessible WebInspectorUIProxy messages into their own interface, WebInspectorBackendProxy. This restricts the interfaces to only what is required for the backend and frontend processes.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.cpp: Added.
(WebKit::WebInspectorBackendProxy::requestOpenLocalInspectorFrontend):
(WebKit::WebInspectorBackendProxy::didClose):
(WebKit::WebInspectorBackendProxy::bringToFront):
(WebKit::WebInspectorBackendProxy::elementSelectionChanged):
(WebKit::WebInspectorBackendProxy::timelineRecordingChanged):
(WebKit::WebInspectorBackendProxy::setDeveloperPreferenceOverride):
(WebKit::WebInspectorBackendProxy::setEmulatedConditions):
(WebKit::WebInspectorBackendProxy::attachAvailabilityChanged):
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.h: Added.
(WebKit::WebInspectorBackendProxy::WebInspectorBackendProxy):
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.messages.in: Copied from Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::WebInspectorUIProxy):
(WebKit::WebInspectorUIProxy::reset):
(WebKit::WebInspectorUIProxy::updateForNewPageProcess):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Inspector/WebInspectorInternal.cpp:
(WebKit::WebInspector::openLocalInspectorFrontend):
(WebKit::WebInspector::closeFrontendConnection):
(WebKit::WebInspector::bringToFront):
(WebKit::WebInspector::elementSelectionChanged):
(WebKit::WebInspector::timelineRecordingChanged):
(WebKit::WebInspector::setDeveloperPreferenceOverride):
(WebKit::WebInspector::setEmulatedConditions):
(WebKit::WebInspector::updateDockingAvailability):

Canonical link: <a href="https://commits.webkit.org/297485@main">https://commits.webkit.org/297485@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ef64598b38f6ecd25d666dd29063757fd46e4ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31541 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22018 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117896 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62106 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40122 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84987 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35672 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25729 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25050 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61738 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95108 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18884 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121176 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38905 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28930 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93858 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39286 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96935 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93682 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23922 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38863 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16650 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34916 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38803 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44298 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38440 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40155 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac72ec646b662bdf6dda81608ad79329da6a79c5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47706

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6a9a8be7fe32481cacd2ff4b7b3eb3fc9dcbb63a. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47706

dcrousso commented:
can you please explain why this change is desired/necessary?

jon-b-apple commented:
> can you please explain why this change is desired/necessary?

It's to resolve a security concern where a compromised WebContent can access the functionality used by Inspector for debugging purposes, since they share the same IPC interface in the UI process. This change splits this into a Backend and Frontend interface and only registers the expected interface for each connection.

burg commented:
a few years ago both backend and frontend had the name "WebInspector" and I renamed WebInspectorProxy to WebInspectorUIProxy. The intent was that [Remote]WebInspectorUI[Proxy] is the Web->UIP interface for local and remote frontends and WebInspector[Proxy] is the Web->UIP interface for backend. But I don't think WebInspectorProxy ever got recreated. and RemoteWebInspectorProxy doesn't exist, since the backend is remote.

I think we can keep the frontend IPC interface as `WebInspectorUIProxy` and move parts of it out to `WebInspectorProxy`. I prefer WebInspectorBackendProxy as it's more descriptive.

jon-b-apple commented:
> a few years ago both backend and frontend had the name "WebInspector" and I renamed WebInspectorProxy to WebInspectorUIProxy. The intent was that [Remote]WebInspectorUI[Proxy] is the Web->UIP interface for local and remote frontends and WebInspector[Proxy] is the Web->UIP interface for backend. But I don't think WebInspectorProxy ever got recreated. and RemoteWebInspectorProxy doesn't exist, since the backend is remote.
> 
> I think we can keep the frontend IPC interface as `WebInspectorUIProxy` and move parts of it out to `WebInspectorProxy`. I prefer WebInspectorBackendProxy as it's more descriptive.

That makes sense, I was trying something like this before when considering splitting up the class. I simplified the patch and went with re-introducing `WebInspectorProxy` as the backend-accessible interface in the UI process, as this feels more consistent with how other interface naming is done. It's still just a thin wrapper around the functionality in `WebInspectorUIProxy` because this class is returned to other parts of WebKit when calling the `inspector()` getter, and I don't yet know if it's safe to remove them (and the referenced data members) from `WebInspectorUIProxy`. As long as they're not accessible via IPC, that should resolve the concern here.

burg commented:
We'll want to audit each message to see whether it's backend or frontend as we split up the interfaces.

This seems in good shape, but in light of my [PR](https://github.com/WebKit/WebKit/pull/47836) that renames InspectorClient to InspectorBackendClient, I'd prefer WebInspectorBackendProxy over WebInspectorProxy.

annevk commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
```diff
@@ -72,14 +73,15 @@ const unsigned WebInspectorUIProxy::initialWindowWidth = 1000;
 
 WebInspectorUIProxy::WebInspectorUIProxy(WebPageProxy& inspectedPage)
-    : m_inspectedPage(inspectedPage)
+    : m_backend(adoptRef(new WebInspectorBackendProxy(*this)))
```
```suggestion
    : m_backend(adoptRef(*new WebInspectorBackendProxy(*this)))
```

annevk commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
```diff
@@ -221,6 +225,8 @@ class WebInspectorUIProxy
     void evaluateInFrontendForTesting(const String&);
 
 private:
+    const RefPtr<WebInspectorBackendProxy> m_backend;
```
```suggestion
    const Ref<WebInspectorBackendProxy> m_backend;
```
I think all of these are `Ref` now. Don't see a need to deviate here, but maybe I'm missing something?

annevk commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
```diff
@@ -72,14 +73,15 @@ const unsigned WebInspectorUIProxy::initialWindowWidth = 1000;
 #endif
 {
-    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorUIProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), *this);
+    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorBackendProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), *m_backend);
```
```suggestion
    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorBackendProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), m_backend.get());
```

annevk commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
```diff
@@ -250,7 +252,7 @@ void WebInspectorUIProxy::updateForNewPageProcess(WebPageProxy& inspectedPage)
     m_inspectedPageIdentifier = inspectedPage.identifier();
 
-    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorUIProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), *this);
+    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorBackendProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), *m_backend);
```
```suggestion
    protectedInspectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebInspectorBackendProxy::messageReceiverName(), m_inspectedPage->webPageIDInMainFrameProcess(), m_backend.get());
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/94d51e19e93bff27bbd4122ec10e0136832af9aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47706

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4446aee1dd8ec7d98732e9cbec51fd6af89b8f36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47706

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/6ef64598b38f6ecd25d666dd29063757fd46e4ec. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47706

jon-b-apple committed:
Split WebInspectorUIProxy into separate interfaces for frontend and backend
rdar://155312814
https://bugs.webkit.org/show_bug.cgi?id=295579

Reviewed by BJ Burg.

First set of changes to split the backend-accessible WebInspectorUIProxy messages into their own interface, WebInspectorBackendProxy. This restricts the interfaces to only what is required for the backend and frontend processes.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.cpp: Added.
(WebKit::WebInspectorBackendProxy::requestOpenLocalInspectorFrontend):
(WebKit::WebInspectorBackendProxy::didClose):
(WebKit::WebInspectorBackendProxy::bringToFront):
(WebKit::WebInspectorBackendProxy::elementSelectionChanged):
(WebKit::WebInspectorBackendProxy::timelineRecordingChanged):
(WebKit::WebInspectorBackendProxy::setDeveloperPreferenceOverride):
(WebKit::WebInspectorBackendProxy::setEmulatedConditions):
(WebKit::WebInspectorBackendProxy::attachAvailabilityChanged):
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.h: Added.
(WebKit::WebInspectorBackendProxy::WebInspectorBackendProxy):
* Source/WebKit/UIProcess/Inspector/WebInspectorBackendProxy.messages.in: Copied from Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::WebInspectorUIProxy):
(WebKit::WebInspectorUIProxy::reset):
(WebKit::WebInspectorUIProxy::updateForNewPageProcess):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Inspector/WebInspectorInternal.cpp:
(WebKit::WebInspector::openLocalInspectorFrontend):
(WebKit::WebInspector::closeFrontendConnection):
(WebKit::WebInspector::bringToFront):
(WebKit::WebInspector::elementSelectionChanged):
(WebKit::WebInspector::timelineRecordingChanged):
(WebKit::WebInspector::setDeveloperPreferenceOverride):
(WebKit::WebInspector::setEmulatedConditions):
(WebKit::WebInspector::updateDockingAvailability):

Canonical link: https://commits.webkit.org/297485@main

webkit-commit-queue commented:
Committed 297485@main (d52237d44775): <https://commits.webkit.org/297485@main>

Reviewed commits have been landed. Closing PR #47706 and removing active labels.
